client/js/Core/componentsandbox.js

```diff
@@ -260,289 +242,360 @@ SCWeb.core.ComponentSandbox.prototype.createViewersForScStructs = function (cont
+                triples.map(triple => withIncomingEdge
+                    ? triple.get("_scene_edge_source") : triple.get("_scene_edge_target")));
+
+            for (let i = 0; i < triples.length; ++i) {
+                const triple = triples[i];
+
+                edgeFromScene = triple.get("_edge_from_scene");
+
+                const sceneEdge = triple.get("_scene_edge");
+                const sceneEdgeHash = sceneEdge.value;
+                const sceneEdgeType = sceneEdgeTypes[i];
+
+                const sceneEdgeElement = withIncomingEdge
+                    ? triple.get("_scene_edge_source")
+                    : triple.get("_scene_edge_target");
+                // if we searched scene structure we'll skip it
+                if (sceneEdgeElement.equal(sceneAddr)) continue;
+
+                if (tracedElements.has(sceneEdgeHash)) continue;
+                tracedElements.add(sceneEdgeHash);
+                nextlevelElements[sceneEdgeHash] = {
+                    type: sceneEdgeType, level: nextLevel, connectorFromScene: edgeFromScene};
+
+                const sceneEdgeElementHash = sceneEdgeElement.value;
+                const sceneEdgeElementType = sceneEdgeElementTypes[i];
+                nextlevelElements[sceneEdgeElementHash] = {type: sceneEdgeElementType, level: nextLevel};
+
+                self.eventStructUpdate({
+                    isAdded: true,
+                    connectorFromScene: edgeFromScene,
+                    sceneElement: sceneEdge,
+                    sceneElementType: sceneEdgeType,
+                    sceneElementStyles: nextLevelStyles,
+                    sceneElementSource: withIncomingEdge ? sceneEdgeElement : mainElement,
+                    sceneElementSourceType: withIncomingEdge ? sceneEdgeElementType : mainElementType,
+                    sceneElementSourceStyles: withIncomingEdge ? nextLevelStyles : levelStyles,
+                    sceneElementTarget: withIncomingEdge ? mainElement : sceneEdgeElement,
+                    sceneElementTargetType: withIncomingEdge ? mainElementType : sceneEdgeElementType,
+                    sceneElementTargetStyles: withIncomingEdge ? levelStyles : nextLevelStyles,
+                });
+            }
+        };
 
-            scTemplate.triple(
-                "src",
-                [sc.ScType.EdgeAccessVarPosPerm, "edgeFromContourToSecondElem"],
-                "secondElem",
-            );
+        let keyElements = self.scAddr ? await verifySceneElement(self.scAddr) : await getSceneMainKeyElements();
+        if (!self.scAddr) self.isSceneWithMainKey = keyElements.length;
+        if (!keyElements.length) keyElements = await getSceneKeyElements();
+        if (self.isSceneWithMainKey || keyElements.length) self.isSceneWithKey = true;
+
+        const elementTypes = await scClient.checkElements(keyElements.map(triple => triple.sceneElement));
+
+        let mainElements = {};
+        for (let i = 0; i < keyElements.length; ++i) {
```

Comment:
what if regular scg will be rendered if there are no key elements?

Reply:
fixed